### PR TITLE
added unit test to guard against native API changes

### LINKED ISSUE #338: Unable to find an entry point named 'ReleaseOrtAllocatorInfo' in DLL 'onnxruntime.dll
**Describe the bug**
The C# library fails with message "System.EntryPointNotFoundException: 'Unable to find an entry point named 'ReleaseOrtAllocatorInfo' in DLL 'onnxruntime.dll'."

**To Reproduce**
See https://github.com/dotnet/machinelearning/issues/2149

### PR DESCRIPTION
**Pull request by jignparm**
Fixes #338 

Adding a unit test to verify that native methods exist. Refactoring CPP code may change the API names, which can go undetected in C# until run time, as seen in the issue below.

https://github.com/dotnet/machinelearning/issues/2106

### CODE REVIEW

**Comment by jignparm** [comment]
> Could you remove OrtCreateMkldnnExecutionProviderFactory and OrtCreateCUDAExecutionProviderFactory from the symbol list?
> Otherwise, people can not build this code without mkldnn enabled.

Removed them from the list.

File: csharp/src/Microsoft.ML.OnnxRuntime/NativeMemoryAllocator.cs [hunk 1]
```diff
@@ -64,7 +64,7 @@ private NativeMemoryAllocatorInfo(IntPtr allocInfo)
 
         private static void Delete(IntPtr nativePtr)
         {
-            NativeMethods.ReleaseOrtAllocatorInfo(nativePtr);
+            NativeMethods.OrtReleaseAllocatorInfo(nativePtr);
```
Comment (eerhardt) [inline csharp/src/Microsoft.ML.OnnxRuntime/NativeMemoryAllocator.cs]: Would it also be beneficial to have a unit test that explicitly tests the `Delete`/`ReleaseHandle` path in the C# library?

File: csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs [hunk 2]
```diff
@@ -170,11 +170,11 @@ internal static class NativeMethods
         [DllImport(nativeLib, CharSet = charSet)]
         public static extern IntPtr /*(OrtStatus*)*/ OrtCreateCUDAExecutionProviderFactory(int device_id, out IntPtr /*(OrtProviderFactoryPtr**)*/ factory);
 
-        [DllImport(nativeLib, CharSet = charSet)]
-        public static extern IntPtr /*(OrtStatus*)*/ OrtCreateNupharExecutionProviderFactory(int device_id, string target_str, out IntPtr /*(OrtProviderFactoryPtr**)*/ factory);
+        //[DllImport(nativeLib, CharSet = charSet)]
+        //public static extern IntPtr /*(OrtStatus*)*/ OrtCreateNupharExecutionProviderFactory(int device_id, string target_str, out IntPtr /*(OrtProviderFactoryPtr**)*/ factory);
```
Comment (eerhardt) [inline csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs]: Can these just be removed instead of commented out?

**Comment by jignparm** [comment]
> Just so you know, I will be removing OrtProviderFactory* stuff from the public headers in my next change. Things will be registered directly into the session options without creating a provider factory first.

Thanks for the heads up. The new (simpler) pattern change will require some refactoring in the C# layer.

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 3]
```diff
@@ -470,6 +473,40 @@ private void TestModelInputFLOAT16()
...
+            "OrtCreateDefaultAllocator","OrtReleaseObject","OrtAllocatorFree","OrtAllocatorGetInfo",
+            "OrtCreateTensorWithDataAsOrtValue","OrtGetTensorMutableData", "OrtReleaseAllocatorInfo",
+            "OrtCastTypeInfoToTensorInfo","OrtGetTensorShapeAndType","OrtGetTensorElementType","OrtGetNumOfDimensions",
+            "OrtGetDimensions","OrtGetTensorShapeElementCount","OrtReleaseValue"};
+
+            var hModule = LoadLibrary(module);
+            foreach (var ep in entryPointNames)
+            {             
+                var x = GetProcAddress(hModule, ep);
+                Assert.False(x == UIntPtr.Zero, $"Entrypoint {ep} not found in module {module}");
```
[the remark this answers is not in the document]
Comment (jignparm) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs]: Asser.NotEqual doesn't allow for a custom message. Having the missing function in the logs would be useful information to have.

File: csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs (hunk 2 above)
Comment (jignparm) [reply csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs]: removed.

File: csharp/src/Microsoft.ML.OnnxRuntime/NativeMemoryAllocator.cs (hunk 1 above)
Comment (jignparm) [reply csharp/src/Microsoft.ML.OnnxRuntime/NativeMemoryAllocator.cs]: Will address this later. Ideally Session.Dispose() should be de-allocating, and this should be a frequently called function. Requires additional refactoring. But agreed would be good to have an explicit unit test.

**Pull request merged**